File: scala/day10_part1_2018.scala
import scala.io.Source
import scala.util.Try

case class Star(var x: Int, var y: Int, vX: Int, vY: Int, var next: Star = null)

object Main extends App {
  val input = Source.fromFile("input.txt").getLines().toList
  val head = new Star(0, 0, 0, 0)
  var tail = head
  val re = """position=<\s*(-?\d+),\s*(-?\d+)>\s*velocity=<\s*(-?\d+),\s*(-?\d+)>""".r
  for (line <- input) {
    line match {
      case re(x, y, vX, vY) =>
        val star = new Star(x.toInt, y.toInt, vX.toInt, vY.toInt)
        tail.next = star
        tail = star
    }
  }

  var smallestT = 0
  var smallestArea = Int.MaxValue
  for (t <- 1 to 100000) {
    var maxX = Int.MinValue
    var maxY = Int.MinValue
    var minX = Int.MaxValue
    var minY = Int.MaxValue

    var temp = head.next
    while (temp != null) {
      val x = temp.x + temp.vX * t
      if (x > maxX) maxX = x
      else if (x < minX) minX = x
      val y = temp.y + temp.vY * t
      if (y > maxY) maxY = y
      else if (y < minY) minY = y
      temp = temp.next
    }

    val lenX = maxX - minX + 1
    val lenY = maxY - minY + 1
    val area = lenX + lenY

    if (area < smallestArea) {
      smallestArea = area
      smallestT = t
    }
  }

  var t = smallestT

  var maxX = Int.MinValue
  var maxY = Int.MinValue
  var minX = Int.MaxValue
  var minY = Int.MaxValue

  var temp = head.next
  while (temp != null) {
    temp.x = temp.x + temp.vX * t
    if (temp.x > maxX) maxX = temp.x
    else if (temp.x < minX) minX = temp.x
    temp.y = temp.y + temp.vY * t
    if (temp.y > maxY) maxY = temp.y
    else if (temp.y < minY) minY = temp.y
    temp = temp.next
  }

  val mapper = Array.ofDim[Boolean](maxY - minY + 1, maxX - minX + 1)

  temp = head.next
  while (temp != null) {
    mapper(temp.y - minY)(temp.x - minX) = true
    temp = temp.next
  }

  for (i <- 0 until mapper.length) {
    for (j <- 0 until mapper(0).length) {
      if (mapper(i)(j)) print("#") else print(" ")
    }
    println()
  }
}

File: c/day10_part1_2018.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

typedef struct Star {
    int x, y, vX, vY;
    struct Star* next;
} Star;

int toInt(char* s) {
    return atoi(s);
}

int main() {
    FILE* file = fopen("input.txt", "r");
    char line[100];
    Star* head = malloc(sizeof(Star));
    Star* tail = head;
    
    while (fgets(line, sizeof(line), file)) {
        int x, y, vX, vY;
        if (sscanf(line, " position=< %d, %d> velocity=< %d, %d>", &x, &y, &vX, &vY) == 4) {
            Star* star = malloc(sizeof(Star));
            star->x = x;
            star->y = y;
            star->vX = vX;
            star->vY = vY;
            tail->next = star;
            tail = star;
        }
    }
    fclose(file);
    
    tail->next = NULL;
    int smallestT = 0, smallestArea = INT_MAX;

    for (int t = 1; t < 100000; t++) {
        int maxX = INT_MIN, maxY = INT_MIN, minX = INT_MAX, minY = INT_MAX;

        for (Star* temp = head->next; temp; temp = temp->next) {
            int x = temp->x + temp->vX * t;
            int y = temp->y + temp->vY * t;
            if (x > maxX) maxX = x;
            if (y > maxY) maxY = y;
            if (x < minX) minX = x;
            if (y < minY) minY = y;
        }

        int area = (maxX - minX + 1) + (maxY - minY + 1);
        if (smallestArea > area) {
            smallestArea = area;
            smallestT = t;
        }
    }

    for (Star* temp = head->next; temp; temp = temp->next) {
        temp->x += temp->vX * smallestT;
        temp->y += temp->vY * smallestT;
    }

    int maxX = INT_MIN, maxY = INT_MIN, minX = INT_MAX, minY = INT_MAX;

    for (Star* temp = head->next; temp; temp = temp->next) {
        if (temp->x > maxX) maxX = temp->x;
        if (temp->y > maxY) maxY = temp->y;
        if (temp->x < minX) minX = temp->x;
        if (temp->y < minY) minY = temp->y;
    }

    char** mapper = malloc((maxY - minY + 1) * sizeof(char*));
    for (int i = 0; i <= maxY - minY; i++) {
        mapper[i] = calloc(maxX - minX + 1, sizeof(char));
        memset(mapper[i], ' ', maxX - minX + 1);
    }

    for (Star* temp = head->next; temp; temp = temp->next) {
        mapper[temp->y - minY][temp->x - minX] = '#';
    }

    for (int i = 0; i <= maxY - minY; i++) {
        printf("%s\n", mapper[i]);
        free(mapper[i]);
    }
    free(mapper);
    
    Star* current = head;
    while (current) {
        Star* next = current->next;
        free(current);
        current = next;
    }
    
    return 0;
}